File: python/Ex_3_2_Solution.py
""" Exercise 3.1: Calculate the trajectory of an observed particle, when
location and orientation of the observer are changing.
The velocities correspond to particle velocities in a central potential.

"""

# Author: Thomas Haslwanter, Date: Nov-2017

import os
from scipy import signal
import skinematics as skin
import numpy as np
import matplotlib.pyplot as plt

def rotate_and_shift(shift=[0,0,0], rotation=0):
    """Get data, and rotate and shift the camera location
    
    Parameters
    ----------
    shift : ndarray, shape (3,)
        Camera translation [Same units as position recordings]
    rotation : float
        Camera rotation [deg]
    """

    # Get the data
    inFile = 'planet_trajectory_2d.txt'
    data = np.loadtxt(inFile)
    
    # Calculate the 3D trajectory
    zData = -data[:,1]*np.tan(30*np.pi/180.)
    data3D = np.column_stack( (data,zData) )
    data3D[:,2] -= 200
    
    # Calculate and plot the 3-D velocity
    vel3D = signal.savgol_filter(data3D, window_length=61, polyorder=3, deriv=1, delta=0.1, axis=0)

    # Show the data
    plt.plot(vel3D[:,0], vel3D[:,1])
    plt.axis('equal')
    plt.title('x/y Velocity')
    plt.xlabel('x')
    plt.ylabel('y')
    plt.show()
    
    # Just to show how to elegantly create 3 subplots
    fig, axs = plt.subplots(3,1, sharex=True)
    for ii in range(3):
        axs[ii].plot(vel3D[:,ii])
        axs[ii].set_ylabel('axis_{0}'.format(ii))
    axs[ii].set_xlabel('Time [pts]')
    axs[0].set_title('Velocities')
    plt.show()
    
    # Shift the location of the camera, using the numpy feature of
    # "broadcasting" to subtract a 3-vector from an Nx3-matrix.
    data_shifted = data3D - np.array(shift)
    
    # Rotate the orientation of the camera
    data_shiftRot = (skin.rotmat.R(0, rotation) @ data_shifted.T).T
    
    # Plot the shifted and rotated trajectory
    outFile = 'shifted_rotated.png'

    plt.plot(data_shiftRot[:,0], data_shiftRot[:,1])
    plt.axhline(0, linestyle='dotted')
    plt.axvline(0, linestyle='dotted')
    plt.xlabel('x')
    plt.ylabel('y')
    plt.title('Shifted & Rotated')

    plt.savefig(outFile, dpi=200)
    print('Data saved to {0}'.format(outFile))
    plt.show()    
    
if __name__=='__main__':
    rotate_and_shift([0, 100, -50], 34)

File: matlab/Ex_3_2_Solution.m
% Exercise 3.2: Calculate the trajectory of an observed particle, when
% location and orientation of the observer are changing.
% The velocities correspond to particle velocities in a central potential.

% Author: Thomas Haslwanter, Date: Jan-2019

shift = [0, 100, -50];
rotation = 34;   % [deg;


% Get the data
inFile = 'planet_trajectory_2d.txt';
data = load(inFile);

% Calculate the 3D trajectory
zData = -data(:,2)*tan(deg2rad(30));
data3D = [data,zData];
data3D(:,3) = data3D(:,3) - 200;

% Calculate and plot the 3-D velocity;
window_length = 61;
polyorder = 3;
deriv = 1;
delta = 0.1;
rate = 1/delta;
vel3D = savgol(data3D, polyorder, window_length, deriv, rate);

% Show the data
plot( vel3D(:,1), vel3D(:,2) );
axis('equal');
title('x/y Velocity');
xlabel('x');
ylabel('y');
shg
disp('Hit any key to continue ...');
pause

% Just to show how to elegantly create 3 subplots
for ii = 1:3
    subplot(3,1,ii);
    plot(vel3D(:,ii));
    ylabel(['axis_' num2str(ii)]);
end
xlabel('Time [pts]');
subplot(311)
title('Velocities');
shg
disp('Hit any key to continue ...');
pause

% Shift the location of the camera
data_shifted = data3D - shift;

% Rotate the orientation of the camera
data_shiftRot = (R('x', rotation) * data_shifted')';

% Plot the shifted and rotated trajectory
outFile = 'shifted_rotated.jpg';

close('all');
plot(data_shiftRot(:,1), data_shiftRot(:,2));
axis('equal');
line(xlim, [0,0],'LineStyle', '--');
line([0,0],ylim, 'LineStyle', '--');
xlabel('x');
ylabel('y');
title('Shifted & Rotated');

print('-djpeg90', outFile)
disp(['Data saved to ' outFile]);

